client/js/widgets/button.js

```diff
@@ -28,6 +28,27 @@ export class Button extends Widget {
       this.domElement.textContent = delta.text;
   }
 
+  applyVariables(field, widgets, variables, problems) {
+    if(Array.isArray(field.applyVariables)) {
+      for(const v of field.applyVariables) {
+        if(v.parameter && v.variable) {
+          field[v.parameter] = (v.index === undefined) ? variables[v.variable] : variables[v.variable][v.index];
+        } else if(v.parameter && v.template) {
+          field[v.parameter] = v.template.replace(/\{([^}]+)\}/g, function(i, key) {
+            return (variables[key] === undefined) ? "" : variables[key];
+          });
+        } else if(v.parameter && v.property) {
+          let w = isValidID(v.widget) ? widgets.get(v.widget) : this;
```

Comment:
isValidID is a function that only exists in `click`. We could replace it with this:
```suggestion
          let w = widgets.has(v.widget) ? widgets.get(v.widget) : this;
```

Reply:
Could move into click() as a sibling of isValidID; but isValidID also can return true for arrays and we don't handle arrays here anyway. Looks like isValidID returns true when given an empty array; not sure if that is an issue for other places it might be called. Not sure what this code would do if given an array of widgets, so just checking widgets.has() seems sensible here.

Reply:
I moved `isValidID()` to be a class member and take `problems` as a 2nd argument. Notice that the current implementation has a recursive call that includes `problems`:

```js
      return !id.map(i=>isValidID(i, problems)).filter(r=>r!==true).length;
```